plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -228,21 +230,22 @@ dischargeCekValue = \case
        or (b) it's needed for an error message.  When we're discharging VBuiltin
        we use arity0 to get the type and term arguments into the right sequence. -}
 
-instance (Closed uni, GShow uni, uni `Everywhere` PrettyConst) => PrettyBy PrettyConfigPlc (CekValue uni) where
+instance (Closed uni, GShow uni, uni `Everywhere` PrettyConst, uni `Everywhere` ExMemoryUsage) =>
+            PrettyBy PrettyConfigPlc (CekValue uni) where
     prettyBy cfg = prettyBy cfg . dischargeCekValue
 
 type instance UniOf (CekValue uni) = uni
 
-instance (Closed uni, uni `Everywhere` ExMemoryUsage) => FromConstant (CekValue uni) where
-    fromConstant = VCon . fromConstant
+instance FromConstant (CekValue uni) where
+    fromConstant = VCon
 
 instance AsConstant (CekValue uni) where
-    asConstant (VCon term) = asConstant term
-    asConstant _           = Nothing
+    asConstant (VCon val) = Just val
+    asConstant _          = Nothing
 
-instance ToExMemory (CekValue uni) where
+instance (Closed uni, uni `Everywhere` ExMemoryUsage) => ToExMemory (CekValue uni) where
     toExMemory = \case
-        VCon t -> termAnn t
+        VCon val -> memoryUsage val
```

Comment:
Could the issue be that `memoryUsage` is defaulting to using the Generics-based `gmemoryUsage` here and that can't see inside the existential so it just assigns it a low cost?  If you have a `Term` then it can see that it really is a big bytestring or whatever and charge you accordingly, but now the details are hidden.

Reply:
But then presumably the currently higher cost comes from annotating the term, and it's not clear why we would annotate this constant with a higher cost, then.

Reply:
> can't see inside the existential so it just assigns it a low cost?

It does know how to calculate the amount of space that a constant takes:

```haskell
instance (Closed uni, uni `Everywhere` ExMemoryUsage) => ExMemoryUsage (Some (ValueOf uni)) where
  memoryUsage (Some (ValueOf uni x)) = bring (Proxy @ExMemoryUsage) uni (memoryUsage x)
```

> If you have a Term then it can see that it really is a big bytestring or whatever and charge you accordingly

That bytestring is stored precisely as a `Some (ValueOf uni)`, i.e. also inside an existential, so there has to be no difference.

> Hmm, maybe not: gmemoryUsage does have a case for Some (ValueOf ...) but I'm not totally sure what that's doing. Someone did open an issue about that: #1861.

That issue is about costing type tags. We currently assume that all type tags take a constant amount of memory, which is probably indeed the case right now, but may change in future. Costing for actual constants should be fine.

I do feel like the diff that you're pointing at is where we have the problem indeed. But I don't understand at the moment what's the culprit of that behaviour.